File: sdk-java/src/test/java/io/superdurable/dex/iwfcompat/InternalChannelWaitingWorkflow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.iwfcompat;

import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.Wait;

final class InternalChannelWaitingWorkflow implements Flow<Integer> {
    final Channel<Integer> channel = Channel.define("waiting-channel", Integer.class);
    private final Step<Integer> start = new Step<Integer>() {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public Wait waitFor(final Context context, final Integer input) {
            return Wait.allOf(channel.forN(2));
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            int output = input;
            for (Integer value : channel.getConditionResults(context)) {
                output += value;
            }
            return StepDecision.gracefulComplete(output);
        }
    };

    @Override
    public StepList<Integer> getSteps() {
        return StepList.startStep(start);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(channel);
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/internal_channel_waiting_workflow.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::sync::LazyLock;

use dex_sdk::{
    Channel, Context, Flow, HandlerResult, PersistenceSchema, Step, StepDecision, StepList, Wait,
};

pub(crate) static CHANNEL: LazyLock<Channel<i32>> =
    LazyLock::new(|| Channel::new("waiting-channel"));

pub(crate) struct InternalChannelWaitingWorkflow {
    start: WaitingStep,
}

impl InternalChannelWaitingWorkflow {
    pub(crate) fn new() -> Self {
        Self { start: WaitingStep }
    }
}

impl Flow for InternalChannelWaitingWorkflow {
    type StartInput = i32;

    fn steps(&self) -> StepList<'_, Self::StartInput> {
        StepList::start(&self.start)
    }

    fn persistence(&self) -> PersistenceSchema {
        PersistenceSchema::new().channel(&CHANNEL)
    }
}

struct WaitingStep;

impl Step for WaitingStep {
    type Input = i32;

    fn wait_for(&self, _context: &mut Context, _input: i32) -> HandlerResult<Wait> {
        Ok(Wait::until(CHANNEL.for_n(2)))
    }

    fn execute(&self, context: &mut Context, input: i32) -> HandlerResult<StepDecision> {
        let output = CHANNEL
            .condition_results(context)?
            .into_iter()
            .fold(input, i32::saturating_add);
        Ok(StepDecision::graceful_complete(output))
    }
}
